### ISSUE TITLE
Upgrade path from django-allauth-2fa to built in Multi-Factor Authentication from #3383 

### ISSUE THREAD

**Issue by violuke** [pennersr/django-allauth#3420]
The new MFA feature is very exciting, thank you! We're using https://github.com/valohai/django-allauth-2fa at the moment and it has served us well for a long time.

Sadly, we can't upgrade allauth yet (due to https://github.com/valohai/django-allauth-2fa/issues/186) but hopefully we can soon.

Going forward, we're wondering if we should switch to the new built-in MFA in allauth to reduce complexity (and we have had compatibility issues in the past). Are there any plans/idea/suggestions for a migration path between the two?

Thank you 🙏

**Comment by pennersr** [comment]
I really cannot give you any advice on whether or not you should switch. Just a few points I ran into when when integrating django-allauth-2fa:

- django-allauth-2fa relies on django-otp. That stores keys in the clear (https://github.com/django-otp/django-otp/issues/67). Fixing that required maintaining a fork of django-otp.

- Ran into some smaller issues, such as: https://github.com/valohai/django-allauth-2fa/pull/131

- Had to adjust some of the flows (e.g. the one of the requirements was to have recovery codes presented on activation).

So, all in all, adding 2FA to the project resulted in a bit more work than I had hoped for.

Now, I want to stress that I do not want to badmouth any of the work the django-allauth-2fa maintainers have done, on the countrary! Trying to bolt-on important flows on an external code base like that of django-allauth is not an easy feat, and it definitely gets the job done.

But, from an allauth point of view, times have changed. A decade ago 2FA was, not that common, and was best explored in a project of its own. But these days, 2FA is basically a core authentication component that every project needs, which is why it was decided to offer support for that out of the box.

As for migrating, that is possible. It is a matter of writing a migration that takes django-allauth-2fa & django-otp records, turning those into `mfa.Authenticator` instances (it supports externally generated recovery codes as well). It would indeed be good to document an example migration of how this can be done...

**Comment by violuke** [comment]
Thanks for the detailed reply; it's very much appreciated.

I'll look to explore migrating the django-allauth-2fa & django-otp records into `mfa.Authenticator` instances and if I make any useful progress, I'll share it here for the benefit of others :+1:

**Comment by pennersr** [comment]
The code below provides a command that migrates the data. **Warning**: This code is experimental.

``` python
import base64

from allauth.mfa.adapter import get_adapter
from allauth.mfa.models import Authenticator
from django.core.management.base import BaseCommand
from django_otp.plugins.otp_static.models import StaticDevice
from django_otp.plugins.otp_totp.models import TOTPDevice


class Command(BaseCommand):
    def handle(self, **options):
        adapter = get_adapter()
        authenticators = []
        for totp in TOTPDevice.objects.filter(confirmed=True).iterator():
            recovery_codes = set()
            for sdevice in StaticDevice.objects.filter(confirmed=True, user_id=totp.user_id).iterator():
                recovery_codes.update(sdevice.token_set.values_list("token", flat=True))
            secret = base64.b32encode(bytes.fromhex(totp.key)).decode("ascii")
            totp_authenticator = Authenticator(
                user_id=totp.user_id,
                type=Authenticator.Type.TOTP,
                data={"secret": adapter.encrypt(secret)},
            )
            authenticators.append(totp_authenticator)
            authenticators.append(
                Authenticator(
                    user_id=totp.user_id,
                    type=Authenticator.Type.RECOVERY_CODES,
                    data={
                        "migrated_codes": [adapter.encrypt(c) for c in recovery_codes],
                    },
                )
            )
        Authenticator.objects.bulk_create(authenticators)
```

**Comment by pennersr** [comment]
Likely we'll need to include the code above in the documentation. I wouldn't want to go as far as actually providing this as a real management command out of the box, as that would make allauth dependent on django-otp.  Any feedback on the command is appreciated, let's wait for positive confirmation first before adding this to the docs.

**Comment by akx** [comment]
I think a management command would be a good idea - the imports for django_otp models can be late and guarded with an `except ImportError: raise ImproperlyConfigured...` sort of thing.

**Comment by pennersr** [comment]
Code that comes out of the box should be properly unit tested and have a certain level of quality. It is not possible to reach that point with catching import errors. I don't want to be responsible for things going wrong when people run this strict a year from now and e.g. django-otp released breaking changes that were not properly tested on this end. Documentation will have to do, then, people are free to run the command as is, or use it in migrations of their own.

**Comment by clokep** [comment]
Would it make more sense to put the management command in django-allauth-2fa? Users can upgrade to that version, run the command then migrate to the native version.

**Comment by akx** [comment]
@clokep That works too.
@pennersr Is the code above something that you'd allow to be adapted for a management command in django-allauth-2fa (with proper attribution)?

**Comment by pennersr** [comment]
@akx Yes, that would be great!

**Comment by pennersr** [comment]
Closed via 0fb75e74

**Comment by pennersr** [comment]
https://codeberg.org/allauth/django-allauth/issues/3420
